gdal/ogr/ogrpolyhedralsurface.cpp

```diff
@@ -0,0 +1,857 @@
+    {
+        oMP.nGeomCount = 0;
+        return OGRERR_NOT_ENOUGH_MEMORY;
+    }
+
+/* -------------------------------------------------------------------- */
+/*      Get the Geoms.                                                  */
+/* -------------------------------------------------------------------- */
+    for( int iGeom = 0; iGeom < oMP.nGeomCount; iGeom++ )
+    {
+        // Parse the polygons
+        unsigned char* pabySubData = pabyData + nDataOffset;
+        if( nSize < 9 && nSize != -1 )
+            return OGRERR_NOT_ENOUGH_DATA;
+
+        OGRwkbGeometryType eSubGeomType;
+        eErr = OGRReadWKBGeometryType( pabySubData, eWkbVariant, &eSubGeomType );
+        if( eErr != OGRERR_NONE )
+            return eErr;
+
+        OGRGeometry* poSubGeom = NULL;
+        eErr = OGRGeometryFactory::createFromWkb( pabySubData, NULL, &poSubGeom, nSize, eWkbVariant );
+
+        if( eErr != OGRERR_NONE )
+        {
+            oMP.nGeomCount = iGeom;
+            delete poSubGeom;
+            return eErr;
+        }
+
+        oMP.papoGeoms[iGeom] = poSubGeom;
+
+        if (oMP.papoGeoms[iGeom]->Is3D())
+            flags |= OGR_G_3D;
+        if (oMP.papoGeoms[iGeom]->IsMeasured())
+            flags |= OGR_G_MEASURED;
+
+        int nSubGeomWkbSize = oMP.papoGeoms[iGeom]->WkbSize();
+        if( nSize != -1 )
+            nSize -= nSubGeomWkbSize;
+
+        nDataOffset += nSubGeomWkbSize;
+    }
+
+    return OGRERR_NONE;
+}
+
+/************************************************************************/
+/*                            exportToWkb()                             */
+/*      Build a well known binary representation of this object.        */
+/************************************************************************/
+
+OGRErr  OGRPolyhedralSurface::exportToWkb ( OGRwkbByteOrder eByteOrder,
+                                            unsigned char * pabyData,
+                                            OGRwkbVariant eWkbVariant ) const
+
+{
+    if( eWkbVariant == wkbVariantOldOgc &&
+        (wkbFlatten(getGeometryType()) == wkbMultiCurve ||
+         wkbFlatten(getGeometryType()) == wkbMultiSurface) ) /* does not make sense for new geometries, so patch it */
+    {
+        eWkbVariant = wkbVariantIso;
+    }
+
+/* -------------------------------------------------------------------- */
+/*      Set the byte order.                                             */
+/* -------------------------------------------------------------------- */
+    pabyData[0] = DB2_V72_UNFIX_BYTE_ORDER((unsigned char) eByteOrder);
+
+/* -------------------------------------------------------------------- */
+/*      Set the geometry feature type, ensuring that 3D flag is         */
+/*      preserved.                                                      */
+/* -------------------------------------------------------------------- */
+    GUInt32 nGType = getGeometryType();
+
+    if ( eWkbVariant == wkbVariantIso )
```

Comment:
We should ignore the actual value of eWkbVariant that is passed to this function, and assume the wkbVariantIso case. (the variant stuff was compatibility with older GDAL / older PostGIS. But we don't need that for the "new" types). So you can simplify all this block